__tests__/server/middleware/pwa/config.spec.js

```diff
@@ -0,0 +1,192 @@
+    })).toMatchObject({
+      enabled: false,
+      escapeHatch: true,
+      noop: false,
+    });
+    expect(setPWAConfig({
+      noop: true,
+    })).toMatchObject({
+      enabled: false,
+      escapeHatch: false,
+      noop: true,
+    });
+  });
+
+  describe('validation', () => {
+    beforeEach(() => {
+      jest.clearAllMocks();
+    });
+    const { error, warn } = console;
+    beforeAll(() => {
+      console.warn = jest.fn();
+      console.error = jest.fn();
+    });
+    afterAll(() => {
+      console.warn = warn;
+      console.error = error;
+    });
+
+    test('invalid configuration object informs the user', () => {
+      expect.assertions(9);
+
+      expect(validatePWAConfig(null)).toEqual(null);
+      expect(console.error).toHaveBeenCalledTimes(1);
+      expect(console.error).toHaveBeenCalledWith('invalid config given to service worker (expected "object")');
+      expect(validatePWAConfig([])).toEqual(null);
+      expect(console.error).toHaveBeenCalledTimes(2);
+      expect(console.error).toHaveBeenCalledWith('invalid config given to service worker (expected "object")');
+      expect(validatePWAConfig(true)).toEqual(null);
+      expect(console.error).toHaveBeenCalledTimes(3);
+      expect(console.error).toHaveBeenCalledWith('invalid config given to service worker (expected "object")');
+    });
+
+    test('invalid configuration keys informs the user and ignores them', () => {
+      expect.assertions(4);
+
+      expect(validatePWAConfig({
+        random: 'key',
+        foo: 'bar',
+      })).toEqual({});
+      expect(console.warn).toHaveBeenCalledTimes(2);
+      expect(console.warn).toHaveBeenCalledWith('supplied configuration key "random" is not a valid property - ignoring');
+      expect(console.warn).toHaveBeenCalledWith('supplied configuration key "foo" is not a valid property - ignoring');
+    });
+
+    test('invalid configuration values for keys informs the user and ignores them', () => {
+      expect.assertions(5);
+
+      expect(validatePWAConfig({
+        enabled: 'true',
+        scope: 42,
+        manifest: '',
+      })).toEqual({});
+      expect(console.warn).toHaveBeenCalledTimes(3);
+      expect(console.warn).toHaveBeenCalledWith('invalid value type given for configuration key "enabled" (expected "Boolean") - ignoring');
+      expect(console.warn).toHaveBeenCalledWith('invalid value type given for configuration key "scope" (expected "String") - ignoring');
+      expect(console.warn).toHaveBeenCalledWith('invalid value type given for configuration key "manifest" (expected "PlainObject") - ignoring');
+    });
+
+    test('valid keys emits no warnings or errors and returns valid configuration', () => {
+      expect.assertions(3);
+
+      const validConfig = {
+        enabled: true,
+        scope: '/',
+        manifest: {},
+      };
+
+      expect(validatePWAConfig(validConfig)).toEqual(validConfig);
+      expect(console.warn).toHaveBeenCalledTimes(0);
```

Comment:
`.not.toHaveBeenCalled()` is a little cleaner #nitpick

Reply:
They have been removed.